### PR TITLE
JWTCreator for basic types

### PR DESCRIPTION
**Pull request by skjolber**
Support for creating tokens using Maps / Lists which must drill down to the base types Boolean, Integer, Long, Double, String and Date.

### CODE REVIEW

**Comment by lbalmaceda** [comment]
This is one of the ideas we proposed back when all this discussion started. I really like what you came up with since it's not exposing any external dependencies in the public API 🎉  Unfortunately, I won't be able to dedicate time to this repo for at least 2 more weeks. I'll come back later

**Comment by skjolber** [comment]
Our use-case is that we are currently migrating to Auth0 from Keycloak and would like to use this library for validation of tokens. So then it is also very convenient to use this library for creating mock tokens in our testing libraries as well - which requires a few deeper structures in the JWT. We are running both providers in parallel so that migration can be done service per service.

**Comment by skjolber** [comment]
Any hope for getting this reviewed in 2018?

**Comment by skjolber** [comment]
See also https://github.com/auth0/java-jwt/pull/278

**Comment by skjolber** [comment]
@lbalmaceda bump. I've added code coverage, the build now passes the codecov quality checks. 

Null keys and values in maps are disallowed, but null items in lists (and arrays) are allowed - fair?

If this gets merged, it would let us remove a test dependency (your competitor's library!) from our projects.

**Comment by skjolber** [comment]
Was any of the alternative PRs merged? If not then do not close.

**Comment by lbalmaceda** [comment]
Hi @skjolber. Apologies for the delay here! We're planning to review this soon. If we happen to request changes, will you be able to address them?

**Comment by skjolber** [comment]
@lbalmaceda yes

File: lib/src/main/java/com/auth0/jwt/JWTCreator.java [hunk 1]
```diff
@@ -290,6 +292,93 @@ public Builder withArrayClaim(String name, Long[] items) throws IllegalArgumentE
...
+                
+                if(entry.getKey() == null || !(entry.getKey() instanceof String)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+        
+        private static boolean validateClaim(List<?> list) {
+            // accept null values in list
```
[the remark this answers is not in the document]
Comment (skjolber) [reply lib/src/main/java/com/auth0/jwt/JWTCreator.java]: The corresponding withArrayClaim(..) does now, so basically it is just the same behaviour.

File: lib/src/main/java/com/auth0/jwt/JWTCreator.java [hunk 2]
```diff
@@ -290,6 +292,93 @@ public Builder withArrayClaim(String name, Long[] items) throws IllegalArgumentE
...
+            }
+            return true;
+        }        
+
+        @SuppressWarnings("unchecked")
+        private static boolean isSupported(Object value) {
+            if(value instanceof List) {
+                return validateClaim((List<?>)value);
+            } else if(value instanceof Map) {
+                return validateClaim((Map<Object, Object>)value);
```
[the remark this answers is not in the document]
Comment (skjolber) [reply lib/src/main/java/com/auth0/jwt/JWTCreator.java]: Technically any map can be passed in by ignoring the type. I'll just cast to the same as the method input.

File: lib/src/main/java/com/auth0/jwt/JWTCreator.java [hunk 3]
```diff
@@ -290,6 +292,93 @@ public Builder withArrayClaim(String name, Long[] items) throws IllegalArgumentE
...
+            assertNonNull(name);
+            // validate map contents
+            if(!validateClaim(items)) {
+                throw new IllegalArgumentException("Expected list containing Map, List, Boolean, Integer, Long, Double, String and Date");
+            }
+            addClaim(name, items);
+            return this;
+        }         
+
+        private static boolean validateClaim(Map<?, Object> map) {
```
[the remark this answers is not in the document]
Comment (skjolber) [reply lib/src/main/java/com/auth0/jwt/JWTCreator.java]: Technically any Map could be passed in. But I'll leave this up to you (and the cast below in isSupportedType(..))

File: lib/src/main/java/com/auth0/jwt/JWTCreator.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (skjolber) [reply lib/src/main/java/com/auth0/jwt/JWTCreator.java]: But are null values so bad? Could leave it up to the caller for value of Map as well.

File: lib/src/main/java/com/auth0/jwt/JWTCreator.java [hunk 4]
```diff
@@ -290,6 +292,93 @@ public Builder withArrayClaim(String name, Long[] items) throws IllegalArgumentE
...
+            } else {
+                return isBasicType(value);
+            }
+        }
+
+        private static boolean isBasicType(Object value) {
+            Class<?> c = value.getClass();
+            
+            if(c.isArray()) {
+                return c == Integer[].class || c == Long[].class || c == String[].class;
```
[the remark this answers is not in the document]
Comment (skjolber) [reply lib/src/main/java/com/auth0/jwt/JWTCreator.java]: Agreed. I think most prefer Lists anyways.

File: lib/src/test/java/com/auth0/jwt/JWTCreatorTest.java [hunk 5]
```diff
@@ -431,4 +436,209 @@ public void shouldAcceptCustomArrayClaimOfTypeLong() throws Exception {
         String[] parts = jwt.split("\\.");
         assertThat(parts[1], is("eyJuYW1lIjpbMSwyLDNdfQ"));
     }
-}
+    
+    @Test
+    public void shouldAcceptCustomClaimOfTypeObject() throws Exception {
```
Comment (lbalmaceda) [inline lib/src/test/java/com/auth0/jwt/JWTCreatorTest.java]: This is testing the "with map" setter. Let's rename it a bit.
```suggestion
    public void shouldAcceptCustomClaimOfTypeMap() throws Exception {
```

File: lib/src/test/java/com/auth0/jwt/JWTCreatorTest.java [hunk 6]
```diff
@@ -431,4 +436,209 @@ public void shouldAcceptCustomArrayClaimOfTypeLong() throws Exception {
...
+        exception.expect(IllegalArgumentException.class);
+
+        JWTCreator.init()
+                 .withClaim("pojo", (Map<String, Object>)data)
+                 .sign(Algorithm.HMAC256("secret"));
+    }
+    
+    @Test
+    public void shouldRefuseCustomListClaimForUnknownListElement() throws Exception{
+    	List<Object> list = Arrays.asList(new UserPojo("Michael", 255));
```
Comment (lbalmaceda) [inline lib/src/test/java/com/auth0/jwt/JWTCreatorTest.java]: Correct. However, if these contents were written into a Map first, the List would have accepted that. Right? Do you think it's worth testing that as well?

Comment (skjolber) [reply lib/src/test/java/com/auth0/jwt/JWTCreatorTest.java]: Adding another test

File: lib/src/main/java/com/auth0/jwt/JWTCreator.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (skjolber) [reply lib/src/main/java/com/auth0/jwt/JWTCreator.java]: Used anon types where possible.

**Pull request merged**